### PR TITLE
add pario for sfc

### PR DESCRIPTION
**Pull request by tsga**
**Description**

**Type of change**

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

**Comment by tsga** [comment]
Note: the parallel write (paranc) appears slower than the non-paranc, especially for write_atm in the "writeincrements_panc" subroutine. Tested for C96E48 and C384E192.

File: src/enkf/controlvec.f90 [hunk 1]
```diff
@@ -215,21 +215,27 @@ subroutine read_control()
 q_ind = getindex(cvars3d, 'q')
 if (q_ind > 0)  allocate(qsat(npts,nlevs,nbackgrounds,nanals_per_iotask))
 if (paranc) then
-   if (nproc == 0) t1 = mpi_wtime()
+   if (nproc == 0) then
```
Comment (ClaraDraper-NOAA) [inline src/enkf/controlvec.f90:218]: Is this change necessary?

File: src/enkf/controlvec.f90 [hunk 2]
```diff
@@ -215,21 +215,27 @@ subroutine read_control()
...
            fgfileprefixes,fgsfcfileprefixes,reducedgrid,grdin,qsat)
 end if
 if (nproc <= ntasks_io-1) then
    if (.not. paranc) then
-      if (nproc == 0) t1 = mpi_wtime()
+      if (nproc == 0) then
+        print *, 'calling readgriddata'
+        t1 = mpi_wtime()
+      end if
       call readgriddata(nanal1(nproc),nanal2(nproc),cvars3d,cvars2d,nc3d,nc2d,clevels,ncdim,nbackgrounds, &
```
Comment (ClaraDraper-NOAA) [inline src/enkf/controlvec.f90:231]: As above.

File: src/enkf/gridio_gfs.f90 [hunk 3]
```diff
@@ -35,6 +35,7 @@ module gridio
 !               a required input for EFSO calculations
 !   2019-03-13  Add precipitation components
 !   2019-07-10  Add convective clouds
+
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90]: delete extra line

File: src/enkf/gridio_gfs.f90 [hunk 4]
```diff
@@ -81,23 +82,29 @@ subroutine readgriddata_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,ntimes, &
   character(len=7) charnanal
 
   real(r_kind) :: kap,kapr,kap1,clip,qi_coef
-
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90]: Add line back.

File: src/enkf/gridio_gfs.f90 [hunk 5]
```diff
@@ -111,11 +118,12 @@ subroutine readgriddata_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,ntimes, &
 
   call set_ncio_file_flags(vars3d, n3d, vars2d, n2d, read_sfc_file, read_atm_file)
 
-   if (read_sfc_file) then
-      print *,'paranc not supported for reading surface files'
-      call mpi_barrier(mpi_comm_world,ierr)
-      call mpi_finalize(ierr)
-   endif
+   ! if (read_sfc_file) then
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90]: remove commented out block.

File: src/enkf/gridio_gfs.f90 [hunk 6]
```diff
@@ -146,20 +155,21 @@ subroutine readgriddata_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,ntimes, &
...
+ if (read_atm_file) then  ! read_atm_file ! 1.18.24 want to test only sfc?
 
   ! loop through times and do the read
   ne = 1
   backgroundloop: do nb=1,ntimes
 
   write(charnanal,'(a3, i3.3)') 'mem', nanal
   filename = trim(adjustl(datapath))//trim(adjustl(fileprefixes(nb)))//trim(charnanal)
-  sfcfilename = trim(adjustl(datapath))//trim(adjustl(filesfcprefixes(nb)))//trim(charnanal)
+  !sfcfilename = trim(adjustl(datapath))//trim(adjustl(filesfcprefixes(nb)))//trim(charnanal)
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90]: This was not actually being used, right?

File: src/enkf/gridio_gfs.f90 [hunk 7]
```diff
@@ -199,302 +209,463 @@ subroutine readgriddata_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,ntimes, &
   ! The following line will be removed after testing
   use_full_hydro = .false.
 
-  if (.not. isinitialized) call init_spec_vars(nlons,nlats,ntrunc,4)
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90]: revert the indent here.

**Review by ClaraDraper-NOAA — CHANGES_REQUESTED** [review]
This is close, thanks. A few things: 
1. remove all added indentation 
2. delete, rather than comment out.

**Comment by tsga** [comment]
- Removed commented out lines
- Fixed indentation to match previous code
- All changed lines now show new/modified code

File: src/enkf/controlvec.f90 [hunk 8]
```diff
@@ -357,7 +363,7 @@ subroutine write_control(no_inflate_flag)
      endif
      deallocate(grdin_mean)
      t2 = mpi_wtime()
-     print *,'time in write_control on root',t2-t1,'secs'
+     print *,'time in write_control parnc on root',t2-t1,'secs'
```
Comment (ClaraDraper-NOAA) [inline src/enkf/controlvec.f90]: change to paranc

File: src/enkf/gridio_gfs.f90 [hunk 9]
```diff
@@ -496,6 +496,143 @@ subroutine readgriddata_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,ntimes, &
...
+   sfcbackgroundloop: do nb=1,ntimes
+   
+   write(charnanal,'(a3, i3.3)') 'mem', nanal
+   sfcfilename = trim(adjustl(datapath))//trim(adjustl(filesfcprefixes(nb)))//trim(charnanal)
+   if (use_gfs_ncio) then
+     dset_sfc = open_dataset(sfcfilename, paropen=.true., mpicomm=iocomms(mem_pe(nproc)))
+     !londim = get_dim(dset,'grid_xt'); nlonsin = londim%len
+     !latdim = get_dim(dset,'grid_yt'); nlatsin = latdim%len
+   else
+     write(6,*)'READGRIDDATA_PNC sfc:  ***FATAL ERROR*** parallel read only supported for netCDF' , ' PROGRAM STOPS'
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90:513]: Check the logic elsewhere, but I think paranc means netcfdf only - so we don't really need this check.

File: src/enkf/gridio_gfs.f90 [hunk 10]
```diff
@@ -4317,6 +4494,9 @@ subroutine writeincrement_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,grdin,no_inflate
      call mpi_bcast(grdin(1,1,nb,1),npts*ndim, mpi_real4, 0, iocomms(mem_pe(nproc)), iret)
   enddo
 
+  if (write_atm_file ) then   ! do only sfc write?
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90]: remove this comment (and ask me about it, if you're not sure).

File: src/enkf/gridio_gfs.f90 [hunk 11]
```diff
@@ -4772,8 +4952,261 @@ subroutine writeincrement_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,grdin,no_inflate
   if (allocated(delzb)) deallocate(delzb)
   if (allocated(psges)) deallocate(psges)
 
+  !TODO: closing file 
+  call nccheck_incr(nf90_close(ncid_out))
```
Comment (ClaraDraper-NOAA) [inline src/enkf/gridio_gfs.f90:4956]: why is this new?

**Review by ClaraDraper-NOAA — APPROVED** [review]
Looks great! Please prepare a PR with your changes for EMC's GSI code base. 

I think it'll be cleanest for you to merge your changes into the latest develop, rather than for me to merge your branch into mine (since I made some changes when mine was reviewed, which will now be merge conflicts).

File: src/enkf/gridio_gfs.f90 (hunk 9 above)
Comment (tsga) [reply src/enkf/gridio_gfs.f90:513]: it is also present in the atm section so keeping it to be consistent

**Pull request closed without merging**